### ISSUE TITLE
containerd task execution sometimes fails with EINTR - interrupted system call

### ISSUE THREAD

**Issue by saj** [concourse/concourse#9027]
Hello,

Since updating to the latest Concourse v7.12.0 release, we have been observing occasional task execution failures.  We are using the containerd runtime on Linux 5.10 on amd64.

> `find or create container on worker REDACTED: starting task: network add: cni net setup: plugin type="loopback" failed (add): interrupted system call`

The failures also sometimes afflict our resource `get` steps.  (I would assume this is more common, though we do not readily observe these failures.)

I do not know whether the problem can be reproduced on demand.  The majority of builds succeed.

I am fairly confident this started with Concourse v7.12.0 but cannot be entirely certain.  We had been using v7.11.2 since February 9, and we fire alerts on some build failures.  I think I would have noticed this problem had it occurred on v7.11.2.

Our Concourse cluster executes on bare metal.  Worker uptime is typically measured in units of weeks or months.  This is all to say that there is little external interference.

What I have not yet been able to understand is that the change to the preemptive scheduler in Go happened a long time ago -- way back in 1.14 -- and most system call interfaces were already [hardened](https://go.dev/doc/go1.15#ospkgos) with `EINTR` retry loops.  It is weird to only see this happen now.

> `{"timestamp":"2024-11-12T23:18:01.379566817Z","level":"error","source":"worker","message":"worker.garden.garden-server.create.failed","data":{"error":"starting task: network add: cni net setup: plugin type=\"loopback\" failed (add): interrupted system call","request":{"Handle":"59cdd84e-498d-48a9-5806-0b795a9ba406","GraceTime":0,"RootFSPath":"raw:///data/volumes/live/2ba00ba4-57a0-4c88-603e-cca16c983fe4/volume","BindMounts":[{"src_path":"/data/volumes/live/37c88e15-bcd4-4ff3-40eb-a75193fa31f3/volume","dst_path":"/scratch","mode":1}],"Network":"","Privileged":false,"Limits":{"bandwidth_limits":{},"cpu_limits":{},"disk_limits":{},"memory_limits":{},"pid_limits":{}}},"session":"1.2.16112"}}`

This `EINTR` may be coming from the CNI plugin process.
(as opposed to some syscall executed by CNI just prior to the CNI plugin)

previously, on Concourse v7.11.2:

```
# /usr/local/concourse/bin/loopback
CNI loopback plugin v1.3.0
CNI protocol versions supported: 0.1.0, 0.2.0, 0.3.0, 0.3.1, 0.4.0, 1.0.0
```

now, on Concourse v7.12.0:

```
# /usr/local/concourse/bin/loopback
CNI loopback plugin v1.6.0
CNI protocol versions supported: 0.1.0, 0.2.0, 0.3.0, 0.3.1, 0.4.0, 1.0.0, 1.1.0
```

---

## Note from Concourse maintainers

We should keep track of this issue and the one's related to it: https://github.com/containernetworking/plugins/issues/1121

From my quick reading of the upstream issues from Docker, it sounds like they wrap the netlink package and reproduce the original behaviour prior to v1.2.1 of the netlink package. We need to investigate if we can do the same or if this is something that containerd or CNI needs to do for us. Not sure what the effects would be if we try to handle this error ourselves.

Docker fixed this for themselves in this PR: https://github.com/moby/moby/pull/48598
k8s issue tracking this problem as well: https://github.com/kubernetes/kubernetes/issues/129562
Useful comment from that thread: https://github.com/kubernetes/kubernetes/issues/129562#issuecomment-2595550251

**Comment by laghoule** [comment]
I can confirm the same issue on Kubernetes deployment:
```text
find or create container on worker concourse-worker-0: starting task: network add: cni net setup: plugin type="loopback" failed (add): interrupted system call
```

Reverting back to 7.11.2 fix the issue.

**Comment by saj** [comment]
Thank you for your reply.  It is helpful to know we did not break it. :)

> This `EINTR` may be coming from the CNI plugin process.

I can probably arrange for some trace code to be deployed to our production cluster.  Will try to isolate the point of failure.

**Comment by saj** [comment]
I think I found the problem.  Filed https://github.com/containernetworking/plugins/issues/1121 with your upstream.

I am going to keep Concourse v7.12.0 -- we really appreciated the IPv6 support! -- but roll us back to CNI plugins v1.5.1 (or thereabouts).

You might want to do the same in your releases.

**Comment by taylorsilva** [comment]
Thanks for digging into this @saj
My guess was that it would have been an issue with containerd/runc ecosystem. Rolling back that version would make sense. Once the pipelines are clear this would make for a good patch update.

**Comment by saj** [comment]
quick follow up report

We reverted back to CNI plugins v1.5.1 on 2024-11-14 UTC.  Everything has been fine since then. 👍 

(We are still using Concourse v7.12.0 with a tailored container image.)

**Comment by tobyhersey-sky** [comment]
We are also seeing this on occasion

**Comment by taylorsilva** [comment]
Not released yet, but https://github.com/concourse/ci/commit/7d38943427fc3e52bd98a75f5e5bc47820ef3ef1 fixes this issue. A 7.12.1 release is imminent: https://github.com/concourse/concourse/issues/9055

**Comment by 9numbernine9** [comment]
Hey @taylorsilva - I just wanted to mention that we've been running `7.12.1` for several days now and haven't seen this issue occur a single time since the upgrade. Thanks for getting this fix in! 🎉

**Comment by taylorsilva** [comment]
Note to self: I deployed the 7.13-rc to ci.concourse-ci.org and all `git` resources that used `HTTPS` url's failed with this error:

```
fatal: unable to access 'https://github.com/concourse/concourse.git/': getaddrinfo() thread failed to start
```

This error is actually coming from curl and if I do `curl -v` it said hostname resolution failed.

Looking at `/etc/resolv.conf`, everything looked fine.

I eventually realized the RC was using the v1.6.2 CNI, which we know has this error. I'm dropping the CNI version back down to v1.5.1 which we know works. Obviously not the same error, but is the only relevant culprit and major difference between 7.12.1 and 7.13-RC

It looks like CNI will get a PR that fixes this soon, so maybe a new release from them in a week or two?

Keep an eye out for this error if it still persists after upgrading CNI past v1.5.1. Maybe something else changed in CNI that caused this DNS lookup issue.

**Comment by saj** [comment]
@taylorsilva your latest post is likely unrelated.  The task (resource get) would not start at all if you hit the race from the OP.

**Comment by taylorsilva** [comment]
Correct, I ended up figuring that out myself after more testing. Will need to figure out the source of that error before the next release.

**Comment by taylorsilva** [comment]
Looks like the upstream issue has been fixed and v1.7.1 of CNI plugins should work for us. I'm going to unpin this in our CI now.

We never were able to reproduce this on demand and I don't recall seeing it happen in our CI. I'm not sure any of our test suites will pick it up. We'll keep an eye on it and leave this issue open for now.

If someone is able to test out 7.13 of Concourse with the 1.7.1 of the CNI plugins and report back here, that would be very helpful 🙏

**Comment by taylorsilva** [comment]
7.13.2 is out with the upgraded CNI plugins. I'm closing this issue as resolved. If someone starts getting this error with the new version then let us know and we'll re-open this issue.
